src/hooks/useWalletsHiddenBalances.ts

```diff
@@ -51,57 +54,65 @@ const useWalletsHiddenBalances = (wallets: AllRainbowWallets): WalletBalanceResu
   );
 
   const calculateHiddenBalanceForAddress = useCallback(
-    (address: Address) => {
+    (address: Address, data?: UserAssetsResult) => {
       const lowerCaseAddress = address.toLowerCase() as Address;
-      const hiddenAssetBalance = getHiddenAssetBalance({ address, nativeCurrency, connectedToHardhat });
+      const hiddenAssetBalance = getHiddenAssetBalance({ address, nativeCurrency, connectedToHardhat, data });
 
-      console.log('calculateHiddenBalanceForAddress', address, hiddenAssetBalance);
-      setHiddenBalances(prev => ({
-        ...prev,
-        [lowerCaseAddress]: hiddenAssetBalance,
-      }));
+      setHiddenBalances(prev => {
+        const newBalance = hiddenAssetBalance;
+        if (!prev[lowerCaseAddress] || !isEqual(prev[lowerCaseAddress], newBalance)) {
+          return {
+            ...prev,
+            [lowerCaseAddress]: newBalance,
+          };
+        }
+        return prev;
+      });
     },
     [nativeCurrency, connectedToHardhat]
   );
 
   useEffect(() => {
-    allAddresses.forEach(calculateHiddenBalanceForAddress);
+    allAddresses.forEach(address => {
+      calculateHiddenBalanceForAddress(address);
+    });
   }, [allAddresses, calculateHiddenBalanceForAddress]);
 
-  // TODO: This is not working as intended. .subscribe() does not trigger when the value changes.
   useEffect(() => {
-    console.log('Setting up subscriptions for addresses:', allAddresses);
+    const assetSubscriptions = allAddresses.map(address => {
```

Comment:
it seems like this is creating a subscription to all query cache events for each account. i think it could be better to just have 1 subscription, and extract the account address from the query key if it's a userAssets query key. that would reduce the number of listener fires by a factor of the # of accounts. 

i'm also wondering if it's possible to have multiple, more focused subscriptions to individual query keys instead of a single subscription to the entire react query cache (for all keys). seems like it's not possible. do you think subscribing to every change in the entire react query cache might have noticeable performance implications?

i think this is another argument for tweaking UserAssetsSync to periodically refetch for all accounts and updating the userAssetsStores

Reply:
yeah I don't think it'll have performance implications if we only have one listener, and I don't think we can subscribe to individual updates unless we use `useQuery` which we can't here since we want to listen for all address keys. I think I'm less worried about this for now, and more worried about the follow-up UserAssetsSync work.

I'll make the change to subscribe to all address keys, but not going to worry about the other change as it's less priority vs. expanding the sync.